ports/espressif/common-hal/analogbufio/BufferedIn.c

```diff
@@ -0,0 +1,283 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * SPDX-FileCopyrightText: Copyright (c) 2023 Milind Movasha
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdio.h>
+#include "common-hal/analogbufio/BufferedIn.h"
+#include "shared-bindings/analogbufio/BufferedIn.h"
+#include "shared-bindings/microcontroller/Pin.h"
+#include "shared/runtime/interrupt_char.h"
+#include "py/runtime.h"
+#include <string.h>
+#include <stdio.h>
+#include "sdkconfig.h"
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "freertos/semphr.h"
+#include "driver/adc.h"
+
+//#define DEBUG_ANALOGBUFIO
+
+#define NUM_SAMPLES_PER_INTERRUPT   256
+#define NUM_ADC_CHANNELS            1
+#define DMA_BUFFER_SIZE	   	        1024
+#define ATTENUATION                 ADC_ATTEN_DB_0
+#define ADC_READ_TIMEOUT_MS         2000
+
+#if defined(CONFIG_IDF_TARGET_ESP32)
+#define ADC_RESULT_BYTE     2
+#define ADC_CONV_LIMIT_EN   1   //For ESP32, this should always be set to 1
+#elif defined(CONFIG_IDF_TARGET_ESP32S2)
+#define ADC_RESULT_BYTE     2
+#define ADC_CONV_LIMIT_EN   0
+#elif defined(CONFIG_IDF_TARGET_ESP32C3) || defined(CONFIG_IDF_TARGET_ESP32H2)
+#define ADC_RESULT_BYTE     4
+#define ADC_CONV_LIMIT_EN   0
+#elif defined(CONFIG_IDF_TARGET_ESP32S3)
+#define ADC_RESULT_BYTE     4
+#define ADC_CONV_LIMIT_EN   0
+#endif
+
+static adc_digi_convert_mode_t convert_mode = ADC_CONV_SINGLE_UNIT_2;
+static adc_digi_output_format_t output_format = ADC_DIGI_OUTPUT_FORMAT_TYPE1;
+static uint8_t adc_channel = 0;
+
+void common_hal_analogbufio_bufferedin_construct(analogbufio_bufferedin_obj_t *self, const mcu_pin_obj_t *pin, uint32_t sample_rate) {
+    uint16_t adc1_chan_mask = 0;
+    uint16_t adc2_chan_mask = 0;
+
+    if( self->pin != NULL) {
+        mp_raise_ValueError(translate("ADC DMA already initialized"));
+    }
```

Comment:
I think this should condition will caught in `shared-bindings/BufferedIn.c`, because it will raise an error for a pin that isn't free. So this check is probably not needed.

Reply:
This check is removed in commit https://github.com/adafruit/circuitpython/pull/7602/commits/ead5751803ae2c770ec5923c5c996cb1f33aa930